src/client/dfuse/ops/readdir.c

```diff
@@ -207,79 +280,272 @@ dfuse_readdir_reset(struct dfuse_readdir_hdl *hdl)
 #define FADP fuse_add_direntry_plus
 #define FAD  fuse_add_direntry
 
-void
-dfuse_cb_readdir(fuse_req_t req, struct dfuse_obj_hdl *oh, size_t size, off_t offset, bool plus)
+int
+dfuse_do_readdir(struct dfuse_projection_info *fs_handle, fuse_req_t req, struct dfuse_obj_hdl *oh,
+		 char *reply_buff, size_t *out_size, off_t offset, bool plus)
 {
-	struct dfuse_projection_info *fs_handle = fuse_req_userdata(req);
-	char                         *reply_buff;
-	off_t                         buff_offset = 0;
-	int                           added       = 0;
-	int                           rc          = 0;
-	bool                          large_fetch = true;
-	struct dfuse_readdir_hdl     *hdl;
-
-	D_ASSERTF(atomic_fetch_add_relaxed(&oh->doh_readir_number, 1) == 0,
-		  "Multiple readdir per handle");
-
-	D_ASSERTF(atomic_fetch_add_relaxed(&oh->doh_ie->ie_readir_number, 1) == 0,
-		  "Multiple readdir per inode");
+	off_t                     buff_offset = 0;
+	int                       added       = 0;
+	int                       rc          = 0;
+	bool                      large_fetch = true;
+	bool                      to_seek     = false;
+	struct dfuse_readdir_hdl *hdl;
+	size_t                    size = *out_size;
+
+	/* Fetch a readdir handle for this operation, this might be shared with other directory
+	 * handles for the same inode.  Only one readdir will happen concurrently for each inode
+	 * however readdir does get called concurrently with releasedir for the same inode so
+	 * protect this section with a spinlock.
+	 */
+	D_SPIN_LOCK(&fs_handle->dpi_info->di_lock);
 
-	if (offset == READDIR_EOD) {
-		oh->doh_kreaddir_finished = true;
-		DFUSE_TRA_DEBUG(oh, "End of directory %#lx", offset);
+	if (oh->doh_rd == NULL) {
+		if (oh->doh_ie->ie_rd_hdl &&
```

Comment:
To be clear, we need to lock whether oh->doh_rd is NULL or not?

Reply:
Good question.  Superficially it would seem not.  Will take a close look tomorrow.